### DISCUSSION TITLE
onPopupCreated()和onBeforePopup()疑问

### DISCUSSION THREAD

**Discussion by L-Super** [CefView/QCefView#241] [Q&A] [answered]
我看到QCefView有[onPopupCreated](https://cefview.github.io/QCefView/zh/docs/reference/QCefView#class_q_cef_view_1aac12eb50ad220fa9ca1cd674ac471a9c)和[onBeforePopup](https://cefview.github.io/QCefView/zh/docs/reference/QCefView#class_q_cef_view_1a3901aeb27b7ee70fd9c8b7e002fb6703)接口，但是函数实现为空，且为槽函数。

所以，我的疑问是，这里是还未实现功能，还是需要继承QCefView去实现对应功能，然后像事件一样被调用？

**Comment by L-Super** [comment]
如果要实现像浏览器一样的多标签页，是否是利用这两个函数进行`addTab()`操作？

**Comment by L-Super** [comment]
@tishion 希望大佬能替我解惑一下

**Reply by tishion** [reply]
文档有误，onPopupCreated函数是需要继承实现的，是为了保证同步调用，如果用slot实现，用户可能使用QueuedConnection来实现，这样就无法保证正确工作了。

这个函数的设计目标是当有popup browser创建的时候给QCefView的使用者一个机会拿到native QWindow，然后可以做一些定制化操作，比如把这个window嵌入一个自定义的Widget里面，然后你可以把这个widget加入tab等，但是这个功能没有经过测试。

**Reply by L-Super** [reply]
感谢回复，之后我会尝试一下这个功能。另外，文档方面，我也做了一些优化，后期希望能汉化一下接口文档，不知作者意见如何？

**Reply by tishion** [reply]
> 感谢回复，之后我会尝试一下这个功能。另外，文档方面，我也做了一些优化，后期希望能汉化一下接口文档，不知作者意见如何？

不胜感谢，文档是通过doxygen提取注释生成xml，然后转成md的，这个过程你可能需要了解一下。

**Reply by L-Super** [reply]
> > 感谢回复，之后我会尝试一下这个功能。另外，文档方面，我也做了一些优化，后期希望能汉化一下接口文档，不知作者意见如何？
> 
> 不胜感谢，文档是通过doxygen提取注释生成xml，然后转成md的，这个过程你可能需要了解一下。

那么，我需要在代码注释里添加中文？提交pr后，它自动生成更新到网站吗？

**Reply by tishion** [reply]
不需要，只需要添加对应的中文MD文档在对应目录即可，需要学习一下[docusaurus](https://docusaurus.io/)

**Reply by L-Super** [reply]
> 文档有误，onPopupCreated函数是需要继承实现的，是为了保证同步调用，如果用slot实现，用户可能使用QueuedConnection来实现，这样就无法保证正确工作了。
> 
> 这个函数的设计目标是当有popup browser创建的时候给QCefView的使用者一个机会拿到native QWindow，然后可以做一些定制化操作，比如把这个window嵌入一个自定义的Widget里面，然后你可以把这个widget加入tab等，但是这个功能没有经过测试。

我在example的基础上，做了尝试，但未达到预期效果，addTab后标签页多了，但是是空白的，而且还会pop一个正常的新窗口。
```cpp
void
CefViewWidget::onPopupCreated(QWindow* wnd)
{
//  emit popWindow(wnd);
  QWidget *widget = QWidget::createWindowContainer(wnd);
  tabWidget->addTab(widget,"new");
}
```
我同时也试着override `onBeforePopup`返回true，但是这样就没有新窗口链接打开动作，更不会pop新窗口。
```cpp
bool
CefViewWidget::onBeforePopup(qint64 frameId,
                             const QString& targetUrl,
                             const QString& targetFrameName,
                             QCefView::CefWindowOpenDisposition,
                             QCefSetting& settings,
                             bool& DisableJavascriptAccess)
{
  return false;
}
```
所以，我不清楚是我实现的有问题，还是此功能不完善导致。

**Comment by tishion** [comment, marked as answer]
我看了一下，文档应该没有错误。
[onBeforePopup](https://cefview.github.io/QCefView/zh/docs/reference/QCefView#class_q_cef_view_1a3901aeb27b7ee70fd9c8b7e002fb6703)
这个方法是在Popup browser还没有创建之前调用，可以在这里修改一些设置/参数

[onPopupCreated](https://cefview.github.io/QCefView/zh/docs/reference/QCefView#class_q_cef_view_1aac12eb50ad220fa9ca1cd674ac471a9c)
这个方法是在Popup browser已经创建并且显示之后调用

如果你需要实现TAB，推荐的方法是在onBeforePopup中直接返回true，取消CEF内置的popup流程，然后拿到url，再创建一个新的QCefView（或者你自己实现的继承自QCefView的类）的对象。

不过这个过程似乎应该直接封装进QCefView更好。
